### PR TITLE
Change flake8 to ruff

### PR DESCRIPTION
**Pull request by nstarman**
### Description

This PR transitions us from flake8 to ruff: a new-but-already-well-established tool for linting. See https://github.com/charliermarsh/ruff for details. In short, ruff is **really** fast. It's easier for the CI and for IDE. I and others (@WilliamJamieson) have been using Ruff for a while and I can attest that, on personal projects using ruff, when I make a change in one part of the codebase ruff has finished checking for inconsistencies throughout the rest of my code by the time I've lifted my finger off of the save shortcut. I highly recommend this quality of life improvement for Astropy.

TLDR:
Replacing flake8 with ruff, which is over 2 orders of magnitude faster and has a lot of auto-fix capabilities.
Ruff is at near feature parity with base flake8, builds in many plugins for native support, and did I mention it's REALLY fast?

Q: ~What happened to the F811?~

~Ruff doesn't support that yet / ever, so those noqa are removed.~

Ruff now support F811, bringing it to full flake8 parity (assuming isort and black are used).

Q: Why get rid of ``yesqa``?

Ruff replaces it! the RUF100 code means ruff will check if a noqa is necessary.

Q: What's happening with the .html stuff in cosmology?

There were weird ascii issues that ruff is correcting.

Q: What's happening with the F401?

Ruff changes a noqa F401 that applies to a block of imports to individual import ignores. Overall, this is an improvement as we can better see which import requires the special treatment.

Q: Does ruff integrate with my IDE?

See https://github.com/charliermarsh/ruff#editor-integrations!
If it doesn't integrate natively, ruff also has a watch mode that reruns on any change (https://github.com/charliermarsh/ruff#usage). If you're thinking that this will take too long to run on every save, allow me to welcome you out of flake8 land, where your computer fan is constantly on max, into this wonderful Eden of ruff.

Q: Are ruff's settings configured on pyproject.toml?

😭. Yes! it's soooo nice to have a linter gladly adopt the future of config files. While I personally like the syntax of setup.cfg more than pyproject.toml, standardization and unification across the ecosystem is vastly more important.

Approvals:
constants: ✅
convolution: ✅ 
coordinates: ✅
cosmology: ✅
io.ascii: ✅ 
io.fits:
io.misc.asdf: ✅ 
nddata:
table: ✅ 
time: ✅ 
units: ✅
visualization: ✅
wcs:

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by pllim** [comment]
Is there no way to just change over the checker without changing the code? Are they not 1-to-1?

File: astropy/modeling/fitting.py
Comment (WilliamJamieson) [inline astropy/modeling/fitting.py:44]: Why is this being removed? This changes the modeling API and is why the tests are failing

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

File: astropy/modeling/fitting.py
Comment (nstarman) [reply astropy/modeling/fitting.py:44]: A good question! I'll do some digging. Might have to add #F401 to the unfixable category.

Comment (WilliamJamieson) [reply astropy/modeling/fitting.py:44]: Looks like they are missing from the `__all__` in that file (which is an oversight). Would adding them there resolve the issue?

Comment (nstarman) [reply astropy/modeling/fitting.py:44]: Worth a try, which I see you've done 👍

File: astropy/units/tests/test_quantity_non_ufuncs.py
Comment (mhvk) [inline astropy/units/tests/test_quantity_non_ufuncs.py:1769]: Just having a very quick look, but I am not in favour of changes like these: this is testing different inputs to functions, and it really helps readability to have the test look like those inputs, i.e., `dict(key=value)` is very much used on purpose here.

I guess I am with @pllim that we should focus here on just replacement for speed reasons, not more stylistic changes.

(Note that I do confirm that python is slightly slower, but I would suggest one's effort should then be to improve python such that it is the same speed rather than change codebases; also it is measured in 10s of ns, so really is not going to matter in all but critical paths.)

File: setup.cfg
Comment (pllim) [inline setup.cfg:170]: 👎 to removing this. My editor actually uses this for "live" PEP 8 checks.

Comment (mhvk) [reply setup.cfg:170]: Same.

Comment (nstarman) [reply setup.cfg:170]: kk. I'll leave it in.
FYI, many editors, e.g. VSCode, also support ruff for live PEP8 checks. Ruff is so fast that it has a live mode that can run in real-time.  https://github.com/charliermarsh/ruff#usage

**Comment by nstarman** [comment]
There's a bit of a conceptual problem with our transfer to ruff. Our current Flake8 enforces different rules in the pre-commit than are in the setup.cfg. While ruff can be similarly configured, I'm having difficulty getting it to not change any files. I'm going to put down this PR for a bit and try to unify our flake8 rules.

**Comment by nstarman** [comment]
With https://github.com/astropy/astropy/pull/14127, the transition to ruff will be a lot easier.

**Review by taldcroft — APPROVED** [review]
Looks fine for io.ascii, time, table.

**Comment by nstarman** [comment]
Will need to be rebased when #14127 is merged.

**Comment by WilliamJamieson** [comment]
> Will need to be rebased when #14127 is merged.

Maybe this should be a draft until that occurs?

File: astropy/units/tests/test_quantity_non_ufuncs.py
Comment (nstarman) [reply astropy/units/tests/test_quantity_non_ufuncs.py]: Done.

**Review by WilliamJamieson — APPROVED** [review]
`io.misc.asdf` changes look good to me

**Comment by nstarman** [comment]
@mhvk @WilliamJamieson @pllim, I restored the E203 ignore and removed the other E1&E2 ignores for flake8. Ruff doesn't use E1&E2 because of black, so that's fine.

**Review by larrybradley — APPROVED** [review]
Looks good for convolution and visualization

**Review by stargaser — APPROVED** [review]
Changes to `constants` look good to me.

**Review by eerovaher — APPROVED** [review]
I can approve the changes to `coordinates`

**Review by mhvk — APPROVED** [review]
units looks all OK now. Nicer to have the changes be small and obvious! Looking forward to trying `ruff` in emacs...

**Comment by mhvk** [comment]
p.s. Some of the `noqa` "fixes" seem to suggest that actual imports should be dropped (e.g., in `nddata`). I think that makes most sense to do outside of this PR, but maybe worth pursuing.

**Comment by nstarman** [comment]
> p.s. Some of the noqa "fixes" seem to suggest that actual imports should be dropped (e.g., in nddata). I think that makes most sense to do outside of this PR

A followup PR SGTM 😆

File: docs/development/codeguide_emacs.rst
Comment (mhvk) [inline docs/development/codeguide_emacs.rst:12]: Does this actually work? I'm checking my emacs setup and it seems not obvious. Also, `ruff` support is currently a feature request for `flycheck`: https://github.com/flycheck/flycheck/issues/1974

Also, `ruff` is not available in Debian or Ubuntu!?!?

Comment (nstarman) [reply docs/development/codeguide_emacs.rst:12]: Thanks for the catch! I saw the issue, and thought it resolved.

For Linux, there should be wheels! https://github.com/charliermarsh/ruff/pull/42

Comment (nstarman) [reply docs/development/codeguide_emacs.rst:12]: About flycheck... we've kept the flake8 settings, so that will continue to function as normal. I'll edit the docs to reflect that. Does the ``ruff —watch`` work?

Comment (mhvk) [reply docs/development/codeguide_emacs.rst:12]: From the flycheck and ruff github issues, it may actually be easy to run it in emacs, but I haven't checked yet. Anyway, this PR should not be held up by that, but the documentation here should not mislead!

**Comment by pllim** [comment]
Hello. My notifications said I got requested multiple times to review this. I'll try to get to this next week... Thanks for your patience!

**Review by pllim — DISMISSED** [review]
Let's revisit later. I don't want this to conflict with the spelling fix or the tox 4 compat PRs that are open right now. I feel like those should get in first.

**Comment by nstarman** [comment]
Also, ruff now supports F811, so I need to rebase and make sure the PR reflect that. It's very good to supports F811, that can prevent hard-to-track-down bugs.

**Comment by WilliamJamieson** [comment]
@nstarman I just noticed that the pre-commit.ci bot was not updating the pin for the ruff pre-commit entry for asdf, see https://github.com/asdf-format/asdf/blob/32b82b8e3e57f7a3ae631dad0ff452bf11cbe8bb/.pre-commit-config.yaml#L58. So we might want to keep that in mind.

**Comment by WilliamJamieson** [comment]
> @nstarman I just noticed that the pre-commit.ci bot was not updating the pin for the ruff pre-commit entry for asdf, see [asdf-format/asdf@`32b82b8`/.pre-commit-config.yaml#L58](https://github.com/asdf-format/asdf/blob/32b82b8e3e57f7a3ae631dad0ff452bf11cbe8bb/.pre-commit-config.yaml#L58). So we might want to keep that in mind.

It maybe because they switched from the keyword `lint` to `ruff` for the job id... So it would fail to install properly with an update.

File: .pre-commit-config.yaml
Comment (WilliamJamieson) [inline .pre-commit-config.yaml:87]: This could be changed to a single line:
```
args: ["--fix"]
```

IMOP this feels cleaner.

Comment (nstarman) [reply .pre-commit-config.yaml:87]: SGTM.

File: .pre-commit-config.yaml
Comment (WilliamJamieson) [inline .pre-commit-config.yaml:83]:
```suggestion
    rev: "v0.0.189"
```
This version includes the `--force-exclude` flag, see charliermarsh/ruff#1295

File: .pre-commit-config.yaml
Comment (WilliamJamieson) [inline .pre-commit-config.yaml:86]:
```suggestion
        args: ["--fix", "--force-exclude"]
```

Add the `--force-exclude` flag.

File: .pre-commit-config.yaml
Comment (WilliamJamieson) [inline .pre-commit-config.yaml:87]:
```suggestion
```
Remove duplicated exclude arguments

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

**Review by WilliamJamieson — APPROVED** [review]

**Comment by nstarman** [comment]
In a followup PR, should we enforce E501? Black ensures it applies to code, but it does not apply to docstrings, comments, nor long strings.

**Comment by WilliamJamieson** [comment]
> In a followup PR, should we enforce E501? Black ensures it applies to code, but it does not apply to docstrings, comments, nor long strings. In #14127 we eliminated the E501 check because, though it was in the setup.cfg configuration, it wasn't enforced in pre-commit, instead relying on maintainers to enforce.

How many errors do we get if we check E501?

**Comment by nstarman** [comment]
> How many errors do we get if we check E501?

A LOT (In a quick test branch I found 184 changed files with 876 additions and 853 deletions). Sadly many of the fixes will be manual. The comments are easily fixed by VSCode and other IDEs' hard wrap functionalities'. Also the docstrings are mostly fixed the same way, with the exception of parameter type descriptions

```
Parameters
——————
arg : if this is super long there's no good way to shorten it.
```

For many of the rest, we have an opportunity to improve the code, e.g. in ``wcs/wcs-api/utils.py`:
```
        s += (('{0:' + str(pixel_dim_width) + 'g}').format(ipix) + '  ' +
                ('{0:' + str(pixel_nam_width) + 's}').format(wcs.pixel_axis_names[ipix] or 'None') + '  ' +
                (" " * 5 + str(None) if pixel_shape[ipix] is None else
                ('{0:' + str(pixel_siz_width) + 'g}').format(pixel_shape[ipix])) + '  ' +
                '{:s}'.format(str(None if wcs.pixel_bounds is None else wcs.pixel_bounds[ipix]) + '\n'))
```
can be rewritten using f-strings as 
```
        s += (
            f"{ipix:{pixel_dim_width}g}  "
            + f"{(wcs.pixel_axis_names[ipix] or 'None'):{pixel_nam_width}s}  "
            + ("     None  " if pixel_shape[ipix] is None
               else f"{pixel_shape[ipix]:{pixel_siz_width}g}  ")
            + "None\n" if wcs.pixel_bounds is None else f"{wcs.pixel_bounds[ipix]:s}\n")
```

**Comment by pllim** [comment]
tox keeps pumping out bug fix, so maybe we don't have to wait for tox 4 compat.

File: astropy/io/ascii/tests/test_cds_header_from_readme.py
Comment (pllim) [inline astropy/io/ascii/tests/test_cds_header_from_readme.py:9]: tl;dr -- Is this undoing `black` changes?

**Comment by pllim** [comment]
> A LOT

Maybe hold off on E501 for now? 😬

File: astropy/io/ascii/tests/test_cds_header_from_readme.py
Comment (nstarman) [reply astropy/io/ascii/tests/test_cds_header_from_readme.py:9]: Sorta —  I think more undoing an `isort` change. ``ruff`` sits on top of `isort` and `black`. The former aggregated these imports, the latter separated them over multiple lines. ``ruff`` assumes a project uses both ``isort`` and ``black`` and adds this further style refinement, disaggregating imports with "errors".

**Comment by nstarman** [comment]
> 

Yeah, definitely for this PR!

**Comment by nstarman** [comment]
> tox keeps pumping out bug fix, so maybe we don't have to wait for tox 4 compat.

🎉 ... I think we have a critical mass of approvals ?...

**Comment by pllim** [comment]
Hold on a bit more. I am trying to get the CI green again...

**Comment by pllim** [comment]
OK CI should be green now in main. Please rebase. Thanks! 

Would be nice to know this didn't accidentally break anything.

**Comment by nstarman** [comment]
🎉

**Pull request merged**
